src/main/java/com/gluonhq/substrate/util/FileOps.java

```diff
@@ -365,4 +373,127 @@ public static void extractFilesFromJar(String extension, Path sourceJar, Path ta
+        }
+        if (Files.isRegularFile(targetDir)) {
+            throw new IOException("Error: " + targetDir + " is not a directory");
+        }
+        if (!Files.exists(targetDir)) {
+            Files.createDirectories(targetDir);
+        }
+        Map<String, String> hashes = new HashMap<>();
+        try (ZipInputStream zis = new ZipInputStream(new FileInputStream(sourceZip.toFile()))) {
+            ZipEntry zipEntry = zis.getNextEntry();
+            while (zipEntry != null) {
+                File destFile = new File(targetDir.toFile(), zipEntry.getName());
+                if (zipEntry.isDirectory()) {
+                    if (!destFile.exists()) {
+                        Files.createDirectories(destFile.toPath());
+                    }
+                } else {
+                    byte[] buffer = new byte[1024];
+                    try (FileOutputStream fos = new FileOutputStream(destFile)) {
+                        int len;
+                        while ((len = zis.read(buffer)) > 0) {
+                            fos.write(buffer, 0, len);
+                        }
+                    }
+                    hashes.put(destFile.getName(), calculateCheckSum(destFile));
+                }
+                zipEntry = zis.getNextEntry();
+            }
+            zis.closeEntry();
+        }
+        return hashes;
+    }
+
+    /**
+     * Downloads a zip file from the specified zipUrl into the targetPath where a file
+     * named zipName will be created.
+     * Note the conventions for the url file (must end with .zip) and the destination file
+     * (must end with .zip)
+     * Once the zip file is downloaded, it will be unpacked into the location that starts at
+     * the targetPath, and is resolved under targetPath/level0/level1/level2
+     * where ether level1 or level2 can be null.
+     * A file with the checksums of all the files in the zip will be generated with extension
+     * ".md5" under the final path: targetPath/level0/level1/level2/level0-level2.md5
+     *
+     * @param zipUrl a string with the location of a zip file, e.g. https://download2.gluonhq.com/substrate/bar/foo.zip
+     * @param targetPath the path where the file zip file will be downloaded, e.g. /opt/bar
+     * @param zipName the name of the file that will be downloaded, e.g. foo.zip
+     * @param level0 the folder under targetPath, not null, e.g. foo1
+     * @param level1 a folder under level0, it can be null, e.g. foo2
+     * @param level2 a folder under level1, it can be null, e.g. foo3
+     *              The zip file will be downloaded into /opt/bar/foo.zip
+     *              The contents of this zip file will be installed into
+     *              /opt/bar/foo1/foo2/foo3, and also the file
+     *               /opt/bar/foo1/foo2/foo3/foo1-foo3.md5 will be created
+     * @throws IOException
+     */
+    public static void processZip(String zipUrl, Path targetPath, String zipName,
```

Comment:
Should the method be named `downloadAndUnzip`?
I might make sense to rethink parameter names:
- `zipUrl -> sourceUrl`
- `targetPath -> destPath` 

Also... does it make sense to define `level` parameters as varargs?

Reply:
Ok, done